sympy/core/basic.py

```diff
@@ -582,9 +582,8 @@ def canonical_variables(self):
         names = {i.name for i in self.free_symbols - set(bound)}
         for b in bound:
             d = next(dums)
-            if b.is_Symbol:
-                while d.name in names:
-                    d = next(dums)
+            while d.name in names:
+                d = next(dums)
```

Comment:
What happens if StopIteration is raise here?

Reply:
It's an infinite generator so I can't imagine a scenario for this.